ninja-core/src/main/java/ninja/ReverseRouter.java

```diff
@@ -0,0 +1,421 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * Reverse routing. Lookup the uri associated with a controller method.
+ * 
+ * @author Joe Lauer (jjlauer)
+ */
+public class ReverseRouter implements WithControllerMethod<Builder> {
+    static private final Logger log = LoggerFactory.getLogger(ReverseRouter.class);
+    
+    static public class Builder {
+        
+        private final String contextPath;
+        private final Route route;
+        private Map<String,String> paths;
+        private Map<String,String> queries;
+        
+        public Builder(String contextPath, Route route) {
+            this.contextPath = contextPath;
+            this.route = route;
+        }
+
+        public Route getRoute() {
+            return route;
+        }
+        
+        /**
+         * Add a parameter as a path replacement. Will validate the path parameter
+         * exists. This method will URL encode the values when building the final
+         * result.
+         * @param name The path parameter name
+         * @param value The path parameter value
+         * @return A reference to this builder
+         * @see #rawPath(java.lang.String, java.lang.Object) 
+         */
+        public Builder path(String name, Object value) {
+            return setPath(name, value, false);
+        }
+        
+        /**
+         * Identical to <code>path</code> except the path parameter value will
+         * NOT be url encoded when building the final url.
+         * @param name The path parameter name
+         * @param value The path parameter value
+         * @return A reference to this builder
+         * @see #path(java.lang.String, java.lang.Object) 
+         */
+        public Builder rawPath(String name, Object value) {
+            return setPath(name, value, true);
+        }
+        
+        private Builder setPath(String name, Object value, boolean raw) {
+            Objects.requireNonNull(name, "name required");
+            Objects.requireNonNull(value, "value required");
+
+            if (route.getParameters() == null || !route.getParameters().containsKey(name)) {
+                throw new IllegalArgumentException("Reverse route " + route.getUri()
+                    + " does not have a path parameter '" + name + "'");
+            }
+            
+            if (this.paths == null) {
+                this.paths = new HashMap<>();
+            }
+            
+            this.paths.put(name, safeValue(value, raw));
+            
+            return this;
+        }
+        
+        /**
+         * Add a parameter as a query string value. This method will URL encode
+         * the values when building the final result.
+         * @param name The query string parameter name
+         * @param value The query string parameter value
+         * @return A reference to this builder
+         * @see #rawQuery(java.lang.String, java.lang.Object) 
+         */
+        public Builder query(String name, Object value) {
```

Comment:
not sure whether queryParam would be more straight forward for the user... (Optional)